move the oauth2 last_used update to the bottom of the request

ryanpetrello opened the pull request:
this avoids acquiring a row lock and holding it for the duration of
transactions (which might include a very slow query)

see: https://github.com/ansible/awx/issues/4694

ryanpetrello commented on awx/main/models/oauth.py:
```diff
@@ -119,9 +118,6 @@ class Meta:
     def is_valid(self, scopes=None):
         valid = super(OAuth2AccessToken, self).is_valid(scopes)
-        if valid:
-            self.last_used = now()
```
This happens *very* early in the request, and obtains an exclusive row lock until the transaction ends:

```python
  /venv/awx/lib64/python3.6/site-packages/django/core/handlers/base.py(113)_get_response()
  /usr/lib64/python3.6/contextlib.py(52)inner()
  /venv/awx/lib64/python3.6/site-packages/django/views/decorators/csrf.py(54)wrapped_view()
  /venv/awx/lib64/python3.6/site-packages/django/views/generic/base.py(71)view()
  /awx_devel/awx/api/generics.py(294)dispatch()
  /venv/awx/lib64/python3.6/site-packages/rest_framework/views.py(478)dispatch()
  /awx_devel/awx/api/generics.py(170)initialize_request()
  /venv/awx/lib64/python3.6/site-packages/rest_framework/request.py(223)user()
  /venv/awx/lib64/python3.6/site-packages/rest_framework/request.py(376)_authenticate()
  /awx_devel/awx
/api/authentication.py(46)authenticate()
  /venv/awx/lib64/python3.6/site-packages/oauth2_provider/contrib/rest_framework/authentication.py(29)authenticate()
  /venv/awx/lib64/python3.6/site-packages/oauth2_provider/oauth2_backends.py(169)verify_request()
  /venv/awx/lib64/python3.6/site-packages/oauthlib/oauth2/rfc6749/endpoints/base.py(87)wrapper()
  /venv/awx/lib64/python3.6/site-packages/oauthlib/oauth2/rfc6749/endpoints/resource.py(75)verify_request()
  /venv/awx/lib64/python3.6/site-packages/oauthlib/oauth2/rfc6749/tokens.py(351)validate_request()
  /venv/awx/lib64/python3.6/site-packages/oauth2_provider/oauth2_validators.py(377)validate_bearer_token()
> /awx_devel/awx/main/models/oauth.py(124)is_valid()
```

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -136,6 +137,12 @@ def process_response(self, request, response):
             request.session.set_expiry(expiry)
             response['Session-Timeout'] = expiry
+
+        token = getattr(request.user, 'oauth_token', None)
```
Middleware _feels_ like the right place to do this, but I'm open to other suggestions.

AlanCoding replied:
4d41e316392bb49f1164c969532b6632a5589d49

matburt replied:
I think middleware is better than in a commit hook which can be obscure and hard to find. The other place may be in the `APIView` generic, but I think middleware is the right place for this.

ryanpetrello replied:
Eh, I actually agree with @AlanCoding, I think `connection.on_commit` would simplify this diff a lot.

AlanCoding replied:
https://docs.djangoproject.com/en/2.2/topics/db/transactions/

> Middleware runs outside of the transaction

Running this outside of the request transaction is the only really important thing here. It looks like either of the solutions will satisfy this.

ryanpetrello committed:
move the oauth2 last_used update to the bottom of the request

this avoids acquiring a row lock (and holding it for the duration of
transactions which include a very slow query)
